test/test_testing.py

```diff
@@ -5232,3 +5232,169 @@ def handler(args: ops.testing.ExecArgs):
         self.assertEqual(args_history[-1].group, "test_group")
         self.assertEqual(args_history[-1].group_id, 4)
         self.assertDictEqual(args_history[-1].environment, {"foo": "hello", "foobar": "barfoo"})
+
+
+class TestActions(unittest.TestCase):
+    def setUp(self):
+        class ActionCharm(ops.CharmBase):
+            def __init__(self, framework: ops.Framework):
+                super().__init__(framework)
+                self.framework.observe(self.on.test_action, self._on_test_action)
+                self.framework.observe(self.on.fail_action, self._on_fail_action)
+                self.framework.observe(self.on.test3_action, self._on_test3_action)
+                self.framework.observe(self.on.bad_results_action, self._on_bad_results_action)
+
+            def _on_test_action(self, event: ops.ActionEvent):
+                pass
+
+            def _on_fail_action(self, event: ops.ActionEvent):
+                event.fail("this will be ignored")
+                event.log("some progress")
+                event.fail("something went wrong")
+                event.log("more progress")
+                event.set_results({"partial": "foo"})
+
+            def _on_test3_action(self, event: ops.ActionEvent):
+                event.log("Step 1")
+                event.set_results({"result1": event.params["foo"]})
+                event.log("Step 2")
+                event.set_results({"result2": event.params.get("bar")})
+
+            def _on_bad_results_action(self, event: ops.ActionEvent):
+                event.set_results({"a": {"b": 1}, "a.b": 2})
+
+        self.harness = ops.testing.Harness(ActionCharm, meta='''
+            name: test
+            ''', actions='''
+            test:
+              description: lorem ipsum
+            fail:
+              description: dolor sit amet
+            test2:
+              description: consectetur adipiscing elit
+              params:
+                foo
+                bar
+              required: [foo]
+              additionalProperties: false
+            test3:
+              description: sed do eiusmod tempor
+              params:
+                foo:
+                  type: string
+                  default: foo-default
+                bar:
+                  type: integer
+            bad-results:
+              description: incididunt ut labore
+            ''')
+        self.harness.begin()
+
+    def test_action_get(self):
+        op = ops.testing._Operation(ops.testing.ActionOutput([], {}), {})
```

Comment:
Hmm, I'm not sure about this. I see what you're doing here: testing the testing-backend separately from the `run_action` method. But I don't think we've generally done that (the implementations are pretty intertwined anyway!) and I think it's less fragile to do just black-box testing here by testing that `run_action` expects/returns the correct things. That way we can change the implementation details / name of `_Operation` / etc without changing the tests.

Do the tests below test all the code paths in any case? If so, I'd say that's good enough.

Happy for push-back if I'm missing something though.

Reply:
Ok, happy to do that. I started out with just the other (higher level) tests, and then added these ones in for completeness. I do agree there is a lot of overlap. There's a little bit that's tested here and not below, but I can merge those together.